src/Functions/icebergTruncate.cpp

```diff
@@ -0,0 +1,181 @@
+                return substr_func->execute(string_arguments, std::make_shared<DataTypeString>(), input_rows_count, false);
+            }
+            else
+            {
+                auto substr_func = FunctionFactory::instance().get("leftUTF8", context)->build(string_arguments);
+                return substr_func->execute(string_arguments, std::make_shared<DataTypeString>(), input_rows_count, false);
+            }
+        }
+        else if (which_truncate.isNativeInteger() || which_truncate.isDecimal())
+        {
+            ColumnsWithTypeAndName modulo_arguments;
+            if (which_truncate.isNativeInteger())
+            {
+                modulo_arguments = {arguments[1], arguments[0]};
+            }
+            else
+            {
+                ColumnPtr decimal_scaled;
+                if (const auto * decimal_type = checkDecimal<Decimal32>(*arguments[1].type))
+                    decimal_scaled = arguments[1].type->createColumnConst(input_rows_count, DecimalField<Decimal32>(value, decimal_type->getScale()));
+                if (const auto * decimal_type = checkDecimal<Decimal64>(*arguments[1].type))
+                    decimal_scaled = arguments[1].type->createColumnConst(input_rows_count, DecimalField<Decimal64>(value, decimal_type->getScale()));
+
+                if (!decimal_scaled)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected decimal data type");
+
+                ColumnWithTypeAndName decimal_scaled_with_type(decimal_scaled, arguments[1].type, "");
+                modulo_arguments = {arguments[1], decimal_scaled_with_type};
+            }
+
+            auto modulo_func = FunctionFactory::instance().get("positiveModulo", context)->build(modulo_arguments);
+            auto modulo_result_type = modulo_func->getResultType();
+            auto modulo_result = modulo_func->execute(modulo_arguments, modulo_result_type, input_rows_count, false);
+            ColumnWithTypeAndName modulo_result_with_type(modulo_result, modulo_result_type, (*modulo_result)[0].dump());
+            auto minus_arguments = {arguments[1], modulo_result_with_type};
+            auto minus_func = FunctionFactory::instance().get("minus", context)->build(minus_arguments);
+            auto minus_result_type = minus_func->getResultType();
+            return minus_func->execute(minus_arguments, minus_result_type, input_rows_count, false);
+        }
+
+        std::unreachable();
+    }
+
+    bool useDefaultImplementationForConstants() const override
+    {
+        return true;
+    }
+
+    DataTypePtr getReturnTypeForDefaultImplementationForDynamic() const override
+    {
+        return std::make_shared<DataTypeString>();
+    }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+};
+
+REGISTER_FUNCTION(icebergTruncate)
+{
+    factory.registerFunction<FunctionIcebergTruncate>();
```

Comment:
We would need to register new functions with documentation - related CI failures.